### PR TITLE
[SPARK-11584][SQL] The attribute of temporay table shows false

### PR DESCRIPTION
**Pull request by dilipbiswal**
The code used to hard-code the temporary attribute to false. Have changed the code to retrieve the attribute from meta store.

Question -
1) I have added a test case in CliSuite as an unit test. When running the tests on jenkins, can this test get executed against older version of hive ? If so, then it will fail as the method to query the temporary attribute is only available after hive 14.

### CODE REVIEW

**Comment by l00357089** [comment]
Hey，I have added your patch to my project and it really works. But now, I find another problem. 
I tried creating a temporary table having the same name with a permanent table, and then show tables, but only one table appears instead of both. Do you have any idea?

**Comment by dilipbiswal** [comment]
@l00357089 Thanks for trying the patch. The behavior you are seeing is consistent with hive. Spark calls hive metastore to get the metadata and therefore sees the same behavior. To confirm , here is what i tried in hive shell -

hive> use dilip;
OK
Time taken: 0.02 seconds
hive> create table t1(c1 int);
OK
Time taken: 0.111 seconds
hive> create temporary table t1(c1 int);
OK
Time taken: 0.088 seconds
hive> show tables;
OK
t1
Time taken: 0.018 seconds, Fetched: 1 row(s)

**Comment by l00357089** [comment]
@dilipbiswal    I know what you mean, but this behavior confuses users and there should be some improvements in spark. I checked the code in consitent  parts and found something. In spark sql, it is overrideCatalog that for temporary tables and hivemetaStoreCatalog for tables from metastore. So, maybe we can firstly register temporary tables to overrideCatalog and then distinguish those from tables in metastore. Actually, I have written some code but it turned out to be the same result as your patch's.  Hope you can solve this problem.

**Comment by l00357089** [comment]
@dilipbiswal Can you find some ways to make sure whether there are two tables having same name ?

**Comment by dilipbiswal** [comment]
@l00357089 I can surely try to work on this if there is a agreement in the community to deviate from hive's semantics. 

Here is the link i came across which describes the semantics.

http://docs.hortonworks.com/HDPDocuments/HDP2/HDP-2.3.0/bk_dataintegration/content/temp-tables.html

Can we open a separate JIRA to handle this since it is a different problem than what is being handled in this JIRA ? 

@rxin @marmbrus Hello Reynold and michael, would like to know your thoughts on this..

**Comment by l00357089** [comment]
@dilipbiswal  I'm a new face here and I 'd like to join you three to learn something.

**Comment by marmbrus** [comment]
I'm not sure we should support the exact semantics of hives temporary tables.  I'd be more inclined to intercept such DDL and convert it to a Spark SQL temporary table (avoiding the metastore completely).  Having two different kinds of temporary table (one materialized the other not unless cached) is weird.

**Comment by dilipbiswal** [comment]
@marmbrus Thanks michael for your input. If we do intercept such DDLS and treat them as Spark SQL  temporary tables, i had a few questions.
1) A command such as SHOW TABLES would now have to go to two different catalogs to get a full listing of temporary and non temporary tables ?
2) Do we allow a permanent and temporary table with the same name ? 
3) If we do, then we need to decide on the scoping rules , right ? If some one does a drop/query table then we need to find out which one to drop/query , right ?

Also michael, did you want me to close this PR ? In this PR, i was simply addressing the defect with missing is_temporary attribute..

Please let me know.

**Comment by marmbrus** [comment]
> 1) A command such as SHOW TABLES would now have to go to two different catalogs to get a full listing of temporary and non temporary tables ?

This is already the case.  We have to go to Hive, and the second catalog is just an in-memory hashtable.

> 2) Do we allow a permanent and temporary table with the same name ? 

We do.

> 3) If we do, then we need to decide on the scoping rules , right ? If some one does a drop/query table then we need to find out which one to drop/query , right ?

Temporary tables will shadow permanent tables.  If you want to access the permanent table you can use `dbName.tableName`.


> Also michael, did you want me to close this PR ? In this PR, i was simply addressing the defect with missing is_temporary attribute..

I'd like to avoid complicating the hive interface unless we are actually going to use this feature.  Given it seems better to just properly support SQL DDL for creating temporary tables I'm not sure it make sense to add it.

**Comment by dilipbiswal** [comment]
@marmbrus  Thanks michael for the answers.

**Comment by yhuai** [comment]
How about we close this PR for now?

**Comment by dilipbiswal** [comment]
@yhuai Hi Yin, i wanted to ask questions to michael and you through this PR on his suggestion to bypass the hive metastore for the temporary table requests. If i close this PR, is there a alternate way to reach either of you ? Please let me know.

**Comment by marmbrus** [comment]
You can continue to comment on a PR after its closed and you can also reopen it if you are the owner of the PR.

**Comment by yhuai** [comment]
What do you mean by `bypass the hive metastore for the temporary table requests`? Sorry, I probably lack of context on it.

**Comment by dilipbiswal** [comment]
@marmbrus Hi Michael, OK.. closing the PR.
@yhuai Yin, here is what michael said in his earlier remark.

I'm not sure we should support the exact semantics of hives temporary tables. I'd be more inclined to intercept such DDL and convert it to a Spark SQL temporary table (avoiding the metastore completely). Having two different kinds of temporary table (one materialized the other not unless cached) is weird.

**Comment by dilipbiswal** [comment]
@marmbrus 
Hi Michael,

I have been studying the code and have a few initial questions.
Looking at the syntax of "create temporary table" for hive..

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Create/Drop/TruncateTable

Here is what is supported today per my understanding.

- create temporay table 
- create temporary table as select
- create temporary table like existing_table

In first two cases, there are quite a few causes that are supported such as
"clustered by", "skewed by", "row format", "stored as" , "location". Per the doc
a temp table can not be partitioned.

Questions :
1. Can you please briefly explain the semantics of "Spark SQL temporary table" ?
   Is it same as when we call dataframe.registerTempTable() ?
2. In hive after creating a temporary table, we can use INSERT.. SELECT to populate
   data. How would we do it when we map it to Spark SQL ?  
3. Michael, i have made some changes in HiveQl.scala:nodeToPlan() where i am intercepting the
   create temp call and would like to create the correct logical plan and need your input.
   - Is the correct place to intercept ?
   - Can you suggest how you think the logical plan should look like ?
   - I tried to insert into a table registered as temp table and get an exception 
     "org.apache.spark.sql.AnalysisException: Inserting into an RDD-based table is not allowed."

Please let me know your thoughts on this.

**Comment by marmbrus** [comment]
This kind of discussion should happen on JIRA.  Briefly, temporary tables are just pointers to a logical plan (similar to a view that is temporary to the session).  I'd don't really think its worth the complexity to also support temporary tables that you can write to unless there are some compelling use cases given that we support standard tables that you can write to.

**Comment by dilipbiswal** [comment]
@marmbrus Thanks as always.. I will move the discussion to JIRA.

**Comment by l00357089** [comment]
@dilipbiswal Hi , Now I have a question and hope you can help me. I added this patch to my project, but I found that one test in hive.execution/SQLQuerySuite failed and error is "No such Table".
Here is the test: test("sanity test for SPARK-6618") .

**Comment by dilipbiswal** [comment]
@l00357089 Hello, i will try this test and get back to you. As you know, based on michael's comments we are not going ahead with this patch. But i will try this suite against my old patch and get back to you shortly.

**Comment by l00357089** [comment]
@dilipbiswal  Hi, I check your patch and found that it is the def getTable you added causes the problem.
As SPARK-6618 shows, there should be correct lock on hive metastore client, so I think there may be a confilct on access to metastore. What do you think?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -491,10 +493,14 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
     result.copy(expectedOutputAttributes = Some(metastoreRelation.output))
   }
 
-  override def getTables(databaseName: Option[String]): Seq[(String, Boolean)] = {
+  override def getTables(databaseName: Option[String]): Seq[(String, Boolean)] = {hive
     val db = databaseName.getOrElse(client.currentDatabase)
+    client.listTables(db).map(tableName => getTable(db, tableName))
```
Comment (l00357089) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @dilipbiswal  Hello, are you still working on this problem? I 'm still interested in this problem and  think it is  this part that causes SPARK-6618 test failed. Cause the test is about concurrent acces to metastore, after client.listTables gets all the tables,  then DROP TABLE action may occur,  and then run getTable will thrown NosuchTable Exception.

Comment (l00357089) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @dilipbiswal  I think that just add a client.withHiveState{}  in line 498 may be ok.

Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @I00357089 Hi...just looked at the code. So are you saying that we should be holding some sort of a lock after the call to listTables ? That seems a little odd to me .. Also i saw other callers to getTable() in HiveMetastoreCatalog and don't see them acquiring any lock. Can you please point me to an example where they acquire lock before accessing catalog artifacts ? Actually i am working on michael's suggestion to by pass hive metastore catalog for temporary tables. I am not actively working on this patch anymore as mike felt we should not change the hive interface much if we are going the other way. Having said that, i am still curious to know about this failure.

Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @l00357089 did you try it and it works ? shouldn't that call go through ClientWrapper which has the enclosing withHiveState ?

Comment (l00357089) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @dilipbiswal  Yes, it works. The test failure does not happen in def client.listTables, but happens in action map(tableName => getTable(db, tableName)). In the procedure of map, a DROP TABLE action may occure, and then getTable you added throw exception.

Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Ah.. i see. Thanks !!

Comment (l00357089) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @dilipbiswal  You are welcome. Hope you can solve this problem quickly based on michael's suggestion.

**Pull request closed without merging**